### PR TITLE
Testing auto instrumentation

### PR DESCRIPTION
**Pull request by tonyredondo**
This `PR` adds autoinstrumentation to the Test funcs.

```go
func TestMain(m *testing.M) {
	log.SetFlags(log.LstdFlags | log.Llongfile)
	nethttp.PatchHttpDefaultClient()
	os.Exit(scopeagent.Run(m, agent.WithSetGlobalTracer()))
}

func TestPass(t *testing.T) {
	//test := scopeagent.StartTest(t)        //Now optional
	//defer test.End()                       //Now optional
	fmt.Println("TestPass finished")
}

func TestFail(t *testing.T) {
	//test := scopeagent.StartTest(t)
	//defer test.End()
	t.Fatal("Failing now")
}
```

And if a `context` is need it in the test:

```go
func TestHTTP(t *testing.T) {
	//test := scopeagent.StartTest(t)
	//defer test.End()

	// Getting the context from the current test
	ctx := scopeagent.GetContextFromTest(t)

	req, err := http.NewRequest("GET", "https://google.com", nil)
	if err != nil {
		t.Fatalf("%+v", err)
	}
	req = req.WithContext(ctx)
	resp, err := http.DefaultClient.Do(req)
	if err != nil {
		t.Fatalf("%+v", err)
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		t.Fatalf("server returned %d status code", resp.StatusCode)
	}
}
```

CODE: https://github.com/undefinedlabs/demo-go/blob/pr-102/main_test.go

SCOPE: https://shared.scope.dev/dashboard/8cee2622-8772-4338-8d72-63653e0e0239/66857b5a-cd46-4b9a-a6b3-438a51b84b4b/default/explore/inspect/fd07a9ec694b506b777bcef2fff2f1ca4c7d8a87/

### CODE REVIEW

**Comment by tonyredondo** [comment]
I did the modifications you tell me @fermayo, this `PR` is ready for review.

File: instrumentation/testing/testing.go [hunk 1]
```diff
@@ -59,7 +77,17 @@ func StartTest(t *testing.T, opts ...Option) *Test {
 
 // Starts a new test with and uses the caller pc info for Name and Suite
 func StartTestFromCaller(t *testing.T, pc uintptr, opts ...Option) *Test {
-	test := &Test{t: t}
+	mutex.Lock()
+	if testPtr, ok := testMap[t]; ok {
+		testPtr.hasEnded = true
+		testPtr.stopCapturingLogs()
+		delete(testMap, testPtr.t)
+	}
```
Comment (fermayo) [inline instrumentation/testing/testing.go]: As discussed, let's try to remove this by:
* Deprecating `test.End()` and always calling our own internal implementation of `end()` using autoinstrumentation
* Making the log capture logic idempotent
* Removing the redundant `delete` statement

File: instrumentation/testing/testing.go [hunk 2]
```diff
@@ -33,11 +34,28 @@ type (
...
+	testMap map[*testing.T]*Test
+
+	intTests *[]testing.InternalTest
+
+	defaultPanicHandler func(*Test)
+)
+
+// Initialize vars
+func init() {
+	mutex = sync.Mutex{}
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply instrumentation/testing/testing.go]: ok

File: instrumentation/testing/testing.go [hunk 3]
```diff
@@ -252,3 +287,53 @@ func StartBenchmark(b *testing.B, pc uintptr, benchFunc func(b *testing.B)) {
...
+			})
+		}
+		// Replace internal tests with new test indirection
+		*intTests = tests
+	}
+}
+
+// Gets the Test struct from testing.T
+func GetTest(t *testing.T) *Test {
+	mutex.Lock()
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply instrumentation/testing/testing.go]: ok

File: instrumentation/testing/testing.go [hunk 4]
```diff
@@ -252,3 +287,53 @@ func StartBenchmark(b *testing.B, pc uintptr, benchFunc func(b *testing.B)) {
...
+	defer mutex.Unlock()
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply instrumentation/testing/testing.go]: ok

File: instrumentation/testing/testing.go [hunk 5]
```diff
@@ -33,11 +34,28 @@ type (
...
 		skipReasonSource string
 		onPanicHandler   func(*Test)
+		hasEnded         bool
 	}
 
 	Option func(*Test)
 )
 
+var (
+	mutex   sync.Mutex
```
[the remark this answers is not in the document]
Comment (tonyredondo) [reply instrumentation/testing/testing.go]: ok

File: instrumentation/testing/testing.go (hunk 1 above)
Comment (tonyredondo) [reply instrumentation/testing/testing.go]: Done without deprecating test.End() only changing the logic to work only in sub tests

File: instrumentation/testing/testing.go [hunk 6]
```diff
@@ -158,30 +225,43 @@ func (test *Test) End() {
 	test.span.Finish()
 }
 
-// Gets the test context
-func (test *Test) Context() context.Context {
-	return test.ctx
+// Gets or create a test struct
+func getOrCreateTest(t *testing.T) (test *Test, exist bool) {
```
Comment (fermayo) [inline instrumentation/testing/testing.go]:
```suggestion
func getOrCreateTest(t *testing.T) (test *Test, exists bool) {
```

**Pull request merged**